src/algorithms/PANOC.jl

```diff
@@ -0,0 +1,250 @@
+end
+
+################################################################################
+# Utility methods
+
+maxit(sol::PANOCIterator) = sol.maxit
+
+converged(sol::PANOCIterator, it) = it > 0 && blockmaxabs(sol.FPR_x)/sol.gamma <= sol.tol
+
+verbose(sol::PANOCIterator) = sol.verbose > 0 
+verbose(sol::PANOCIterator, it) = sol.verbose > 0 && (sol.verbose == 2 ? true : (it == 1 || it%sol.verbose_freq == 0))
+
+function display(sol::PANOCIterator)
+	@printf("%6s | %10s | %10s | %10s | %10s |\n ", "it", "gamma", "fpr", "tau", "FBE")
+	@printf("------|------------|------------|------------|------------|\n")
+end
+function display(sol::PANOCIterator, it)
+    @printf("%6d | %7.4e | %7.4e | %7.4e | %7.4e | \n", it, sol.gamma, blockmaxabs(sol.FPR_x)/sol.gamma, sol.tau, sol.FBE_x)
+end
+
+function Base.show(io::IO, sol::PANOCIterator)
+	println(io, "PANOC" )
+	println(io, "fpr        : $(blockmaxabs(sol.FPR_x))")
+	println(io, "gamma      : $(sol.gamma)")
+	println(io, "tau        : $(sol.tau)")
+	print(  io, "FBE        : $(sol.FBE_x)")
+end
+
+################################################################################
+# Initialization
+
+function initialize!(sol::PANOCIterator)
+
+    # compute first forward-backward step here
+    A_mul_B!(sol.Aqx, sol.Aq, sol.x)
+    sol.fq_Aqx = gradient!(sol.gradfq_Aqx, sol.fq, sol.Aqx)
+    A_mul_B!(sol.Asx, sol.As, sol.x)
+    sol.fs_Asx = gradient!(sol.gradfs_Asx, sol.fs, sol.Asx)
+    blockaxpy!(sol.At_gradf_Ax, sol.As'*sol.gradfs_Asx, 1.0, sol.Aq'*sol.gradfq_Aqx)
+    sol.f_Ax = sol.fs_Asx + sol.fq_Aqx
+
+    if sol.gamma <= 0.0 # estimate L in this case, and set gamma = 1/L
+        # this part should be as follows:
+        # 1) if adaptive = false and only fq is present then L is "accurate"
+        # 2) otherwise L is "inaccurate" and set adaptive = true
+        # TODO: implement case 1), now 2) is always performed
+        xeps = sol.x .+ sqrt(eps())
+        Aqxeps = sol.Aq*xeps
+        gradfq_Aqxeps, = gradient(sol.fq, Aqxeps)
+        Asxeps = sol.As*xeps
+        gradfs_Asxeps, = gradient(sol.fs, Asxeps)
+        At_gradf_Axeps = sol.As'*gradfs_Asxeps .+ sol.Aq'*gradfq_Aqxeps
+        L = blockvecnorm(sol.At_gradf_Ax .- At_gradf_Axeps)/(sqrt(eps()*blocklength(xeps)))
+        sol.adaptive = true
+        # in both cases set gamma = 1/L
+        sol.gamma = sol.alpha/L
+    end
+
+    blockaxpy!(sol.y, sol.x, -sol.gamma, sol.At_gradf_Ax)
+    sol.g_xbar = prox!(sol.xbar, sol.g, sol.y, sol.gamma)
+    blockaxpy!(sol.FPR_x, sol.x, -1.0, sol.xbar)
+
+    normFPR_x = blockvecnorm(sol.FPR_x)
+    sol.FBE_x = sol.f_Ax - blockvecdot(sol.At_gradf_Ax, sol.FPR_x) + 0.5/sol.gamma*normFPR_x^2 + sol.g_xbar
+
+
+    if sol.adaptive
+	    Aqxbar = sol.Aq*sol.xbar
+	    gradfq_Aqxbar, fq_Aqxbar = gradient(sol.fq, Aqxbar)
+	    Asxbar = sol.As*sol.xbar
+	    gradfs_Asxbar, fs_Asxbar = gradient(sol.fs, Asxbar)
+	    f_Axbar = fs_Asxbar + fq_Aqxbar
```

Comment:
The five lines above are duplicates of what’s at the bottom of the loop. Can avoid the repetition?